starfish/pipeline/features/codebook.py

```diff
@@ -0,0 +1,542 @@
+import functools
+import json
+from typing import Any, Dict, List, Optional, Sequence, Set
+import uuid
+
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+from starfish.constants import Indices, AugmentedEnum
+from starfish.pipeline.features.intensity_table import IntensityTable
+
+
+class Codebook(xr.DataArray):
```

Comment:
I love this. Only thing I'm struggling with, is why the Codebook exposes decoders? It feels cleaner to me to expose a Decoder class that takes a Codebook and an IntensityTable and returns an IntensityTable. This allows the Codebook class to be method agnostic. For example, StarMAP, MARCOFISH, and DARTFISH all use a variant of Codebook.decode_per_hyb_max() and MERFISH and SEQFISH use a variant of Codebook.decode_euclidean. The variance across decoders arises from the error correction, the desired normalizaton (of codewords/intensities) and the desired quality metrics. I think this method specific complexity will obfuscate the Codebook class, which is so elegant. How about moving this complexity to seperate Decoder objects as mentioned above? Or am I missing something?

Reply:
> I think this method specific complexity will obfuscate the Codebook class

I think each of the things you described should be parameters for the `decode_per_hyb_max` and `decode_euclidean` methods. Then, the `Decoder(PipelineComponent)` can call the methods with assay-specific parameters. 

I don't have strong feelings about whether the `Codebook` or `Decoder` exposes the two decode methods above (but I definitely don't want more than those two `decode` methods, unless we think it's unavoidable).

Reply:
Yeah, unless there's a compelling reason not to move the decoders into their own decoder class, I say we do it. But we can do that in another PR.